src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -189,6 +189,12 @@ private static void ReadCollectionItem(ODataReader reader, Stack<ODataItemBase>
                     itemsStack.Push(deltaResourceSetWrapper);
                     break;
 
+                case ODataReaderState.NestedProperty:
+                    Contract.Assert(itemsStack.Count > 0, "The nested property info should be a non-null primitive value within resource wrapper.");
+                    ODataResourceWrapper resourceParentWrapper = (ODataResourceWrapper)itemsStack.Peek();
```

Comment:
This assumes that the only time we would get a NestedProperty is when reading a property info with no value. Is that guaranteed to be true? Might the PropertyInfo be an ODataProperty (i.e., have a value)? Now or in the future?  In that case would it also be added to ODataResource.Properties?